jack-sys/build.rs

```diff
@@ -1,14 +1,2010 @@
+    //     free_description_itself: ::libc::c_int,
+    // ) -> ();
+    Function {
+        name: "jack_free_description",
+        args: &[
+            ("desc", "*mut jack_description_t"),
+            ("free_description_itself", "::libc::c_int"),
+        ],
+        ret: "()",
+        flags: FunctionFlags::NONE,
+    },
+    // pub fn jack_get_properties(subject: jack_uuid_t, desc: *mut jack_description_t) -> ::libc::c_int;
+    Function {
+        name: "jack_get_properties",
+        args: &[
+            ("subject", "jack_uuid_t"),
+            ("desc", "*mut jack_description_t"),
+        ],
+        ret: "::libc::c_int",
+        flags: FunctionFlags::NONE,
+    },
+    // pub fn jack_get_all_properties(descs: *mut *mut jack_description_t) -> ::libc::c_int;
+    Function {
+        name: "jack_get_all_properties",
+        args: &[("descs", "*mut *mut jack_description_t")],
+        ret: "::libc::c_int",
+        flags: FunctionFlags::NONE,
+    },
+    // pub fn jack_remove_property(
+    //     client: *mut jack_client_t,
+    //     subject: jack_uuid_t,
+    //     key: *const ::libc::c_char,
+    // ) -> ::libc::c_int;
+    Function {
+        name: "jack_remove_property",
+        args: &[
+            ("client", "*mut jack_client_t"),
+            ("subject", "jack_uuid_t"),
+            ("key", "*const ::libc::c_char"),
+        ],
+        ret: "::libc::c_int",
+        flags: FunctionFlags::NONE,
+    },
+    // pub fn jack_remove_properties(client: *mut jack_client_t, subject: jack_uuid_t) -> ::libc::c_int;
+    Function {
+        name: "jack_remove_properties",
+        args: &[("client", "*mut jack_client_t"), ("subject", "jack_uuid_t")],
+        ret: "::libc::c_int",
+        flags: FunctionFlags::NONE,
+    },
+    // pub fn jack_remove_all_properties(client: *mut jack_client_t) -> ::libc::c_int;
+    Function {
+        name: "jack_remove_all_properties",
+        args: &[("client", "*mut jack_client_t")],
+        ret: "::libc::c_int",
+        flags: FunctionFlags::NONE,
+    },
+    // pub fn jack_set_property_change_callback(
+    //     client: *mut jack_client_t,
+    //     callback: JackPropertyChangeCallback,
+    //     arg: *mut ::libc::c_void,
+    // ) -> ::libc::c_int;
+    Function {
+        name: "jack_set_property_change_callback",
+        args: &[
+            ("client", "*mut jack_client_t"),
+            ("callback", "JackPropertyChangeCallback"),
+            ("arg", "*mut ::libc::c_void"),
+        ],
+        ret: "::libc::c_int",
+        flags: FunctionFlags::NONE,
+    },
+    // pub fn jack_get_internal_client_name(
+    //     client: *mut jack_client_t,
+    //     intclient: jack_intclient_t,
+    // ) -> *mut ::libc::c_char;
+    Function {
+        name: "jack_get_internal_client_name",
+        args: &[
+            ("client", "*mut jack_client_t"),
+            ("intclient", "jack_intclient_t"),
+        ],
+        ret: "*mut ::libc::c_char",
+        flags: FunctionFlags::WEAK,
```

Comment:
`jack_get_internal_client_name` is not marked as a weak function [in the C header](https://github.com/jackaudio/jack2/blob/852ba2b8ebc944865e75d8eb5d4550d408772f3b/common/jack/intclient.h#L45).

Reply:
If you really think the functions marked as weak in the JACK C headers should be handled specially, then another alternative solution could be adding the capability to handle these to dlib. But `jack_get_internal_client_name` isn't marked as a weak function so I think that's tangential to the error you encountered.

Reply:
Weak is probably the wrong term, maybe something more like optional. Sorry about any disagreements, I hope to address them sometime later, some key takeaways:

1. To me, failing on Linux without specifying feature flags is a huge failure of the current library and I treated it with some sort of urgency. I think in your opinion, I have some niche setup. I disagree. And either way, niche setups are common in the Linux niche.
2. Due to strongly believing in point 1, I did push this out somewhat in a hurry. The options I saw for myself are to push it out in a hurry because I believe in point 1, sit on the current issues which reflect poorly on my library and address them when there is time, spend more time working on this at the expense of my mental health and life.
3. The ability to `link` the library was removed and I may be undervaluing this functionality. The justification I saw was that package maintainers recommend this, but it didn't click with me as there are many cases where applications want to ship JACK as optional anyways.
4. There is some concern that I have a half baked codegen, fair enough. However, I would note that the exposed interface is practically identical to jack_sys before dlopen was rolled out. This is a pretty successful migration IMO and something done the right way can replace what I did but provide the right interface.

Something else to note, I do this in my own free time (though thankfully I have a $5 monthly sponsorship). My day job is in no way similar to this. What this means is that I'm driven by the joy of writing software and the pride in producing what I believe to be quality (given the constraints). Not addressing point 1 when I feel its easy is really demotivating.

Reply:
>  I think in your opinion, I have some niche setup. I disagree.

Yes, I think you do. Ubuntu does not recommend Pipewire for audio yet. Technically the package exists but it still requires manual setup to use. I personally haven't heard of anyone else using Pipewire for audio on Ubuntu or Debian using the distro packages (besides myself when I briefly ran Mobian on my Pinephone Pro before I got Fedora on it); I think people who are using Pipewire for audio on these distros are generally using [this PPA](https://pipewire-debian.github.io/pipewire-debian/) which keeps up to date with upstream Pipewire.

I say this not to dismiss the issue you ran into, but I think you overestimated its impact because you ran into it on your own machine. Recall that nobody said anything about these bindings not working with Pipewire until I did in #142 and fixed it months later in #154. It's good to take pride in your work, but you don't need to let that stress you out. If you want help maintaining this, I'd be happy to help. I doubt anyone would think it would reflect poorly on you if you didn't immediately have a solution to the problem you encountered, especially considering nobody else has reported it. By contrast, as a developer also depending on this library, it would have looked better to me to fully investigate the source of the problem and take the time to consider the pros and cons of various solutions before rushing a +2,794 −2,031 major rewrite of jack-sys that removes an important feature.

I'm still confused about the conditions you encountered the [error](https://github.com/RustAudio/rust-jack/pull/172#issuecomment-1094139332). The error looks like it happened at runtime when you used the `dlopen` feature, which was enabled by default, but you said it occurred with `default-features = false`. If it only occurs with the `dlopen` feature, a simple solution would be disabling that by default. On further thought, I think this may be a good idea in general regardless of the issue with `jack_get_internal_client_name`. JACK is most commonly used on Linux, and it's helpful for Linux packagers to link at build time, so I think it would be good to have the default work well for Linux packaging. Applications that want to use these bindings on Windows and macOS could enable the `dlopen` feature for those operating systems.

> I would note that the exposed interface is practically identical to jack_sys before dlopen was rolled out. This is a pretty successful migration IMO

Yes, I am glad to have the safe Rust abstractions that don't leak implementation details of jack-sys. :smile:

Reply:
I am resuming working on my project which uses this library and ran into this issue again when my collaborator bumped the jack crate to 0.10.0 without knowing about this issue. I hope we can reach a consensus on a long term solution. 
I do not think writing macros just for this crate as proposed in #174 is a good idea because that would pretty much be reimplementing dlib. Ubuntu 21.10 has been outdated by 22.04 for several months now, so I don't think it's worthwhile to go out of the way to work around an upstream bug that has already been fixed. I think the best way forward would be to revert this pull request. Unfortunately that's not trivially simple anymore since you merged #175 after this. I could do the work of reimplementing #175 after reverting to before this PR.